include/ck/tensor_operation/gpu/grid/gridwise_layernorm.hpp

```diff
@@ -0,0 +1,368 @@
+                                                                     OutDstVectorSize,
+                                                                     InMemoryDataOperationEnum::Set,
+                                                                     1,
+                                                                     true>(
+            out_grid_desc_m_k,
+            make_multi_index(blkgroup_id * M_BlockTileSize + thread_m_cluster_id * MThreadSliceSize,
+                             block_local_id * reduceSizePerBlock +
+                                 thread_k_cluster_id * KThreadSliceSize),
+            PassThroughOp{});
+
+        constexpr auto in_thread_copy_fwd_step =
+            make_multi_index(0, SweepOnce ? 0 : K_BlockTileSize);
+        constexpr auto in_thread_copy_bwd_step =
+            make_multi_index(0, SweepOnce ? 0 : -K_BlockTileSize);
+
+        const auto in_global_val_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_x_global, in_grid_desc_m_k.GetElementSpaceSize());
+
+        const auto gamma_global_val_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_gamma_global, gamma_grid_desc_m_k.GetElementSpaceSize());
+
+        const auto beta_global_val_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_beta_global, beta_grid_desc_m_k.GetElementSpaceSize());
+
+        // E(x), E[x^2], var(x)
+        int reduce_length    = in_grid_desc_m_k.GetLength(I1);
+        index_t reducedTiles = 0;
+        do
+        {
+            threadwise_x_load.Run(in_grid_desc_m_k,
+                                  in_global_val_buf,
+                                  thread_buffer_desc,
+                                  make_tuple(I0, I0),
+                                  in_thread_buf);
+
+            static_for<0, MThreadSliceSize, 1>{}([&](auto iM) {
+                static_for<0, KThreadSliceSize, 1>{}([&](auto iK) {
+                    constexpr auto offset = thread_buffer_desc.CalculateOffset(make_tuple(iM, iK));
+                    in_square_thread_buf(Number<offset>{}) =
+                        in_thread_buf(Number<offset>{}) * in_thread_buf(Number<offset>{});
+                });
+            });
+
+            ThreadwiseSumReduce::Reduce(in_thread_buf, mean_thread_buf);
+            ThreadwiseSumReduce::Reduce(in_square_thread_buf, mean_square_thread_buf);
+
+            threadwise_x_load.MoveSrcSliceWindow(in_grid_desc_m_k, in_thread_copy_fwd_step);
+
+            ++reducedTiles;
+        } while(reducedTiles < num_k_block_tile_iteration);
+
+        static_for<0, MThreadSliceSize, 1>{}([&](auto I) {
+            BlockwiseSumReduce::Reduce(reduce_work_buf, mean_thread_buf(I));
+            mean_thread_buf(I) = mean_thread_buf(I) / reduce_length;
+
+            BlockwiseSumReduce::Reduce(reduce_work_buf, mean_square_thread_buf(I));
+            mean_square_thread_buf(I) = mean_square_thread_buf(I) / reduce_length;
```

Comment:
reduce_length is obtained from the input length which may have been padded. Have you checked if having problem size not divisible by tile size produces the right answer?

Reply:
6d3ad8cd1abfca267d2597380b35d0d2f9250c5b